controller/state.go

```diff
@@ -454,6 +471,47 @@ func (m *appStateManager) CompareAppState(app *v1alpha1.Application, project *ap
 		conditions = append(conditions, appv1.ApplicationCondition{Type: v1alpha1.ApplicationConditionComparisonError, Message: err.Error(), LastTransitionTime: &now})
 	}
 
+	// Git has already performed the signature verification via its GPG interface, and the result is available
+	// in the manifest info received from the repository server. We now need to form our oppinion about the result
+	// and stop processing if we do not agree about the outcome.
+	if gpg.IsGPGEnabled() && verifySignature && manifestInfo != nil {
+		// We need to have some data in the verificatin result to parse, otherwise there was no signature
+		if manifestInfo.VerifyResult != "" {
+			verifyResult, err := gpg.ParseGitCommitVerification(manifestInfo.VerifyResult)
+			if err != nil {
+				conditions = append(conditions, v1alpha1.ApplicationCondition{Type: v1alpha1.ApplicationConditionComparisonError, Message: err.Error(), LastTransitionTime: &now})
+				log.Errorf("Error while verifying git commit for revision %s: %s", revision, err.Error())
+			} else {
+				switch verifyResult.Result {
+				case gpg.VerifyResultGood:
+					// This is the only case we allow to sync to, but we need to make sure signing key is allowed
+					validKey := false
+					for _, k := range project.Spec.SignatureKeys {
+						if gpg.KeyID(k.KeyID) == gpg.KeyID(verifyResult.KeyID) && gpg.KeyID(k.KeyID) != "" {
```

Comment:
Semantics of `gpg.KeyID(str string) string` is to return the short KeyID for a GnuPG key `str` from either a long key ID (aka fingerprint) or short key ID (`str` is returned). If `str` is neither long key ID nor short key ID, the empty string `""` is returned, therefore the check on `gpg.KeyID(k.KeyID) != ""` is required to ensure a valid key ID.

I was first thinking to have it return `string, error`, but then again, it would have complicated things.